Python/ceval.c

```diff
@@ -2955,97 +2954,57 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
-               In the first case, we must call
-                 EXIT(None, None, None)
-               otherwise we must call
-                 EXIT(TOP, SECOND, THIRD)
-
-               In the first case, we remove EXIT from the
-               stack, leaving TOP, and push TOP on the stack.
-               Otherwise we shift the bottom 3 values of the
-               stack down, replace the empty spot with NULL, and push
-               None on the stack.
-
-               Finally we push the result of the call.
-            */
-            PyObject *stack[3];
-            PyObject *exit_func;
-            PyObject *exc, *val, *tb, *res;
+            /* (TOP, SECOND, THIRD) = exc_info()
+               (FOURTH, FIFTH, SIXTH) = previous exception for EXCEPT_HANDLER
+               SEVENTH = EXIT, the context.__exit__ or context.__aexit__
+                  bound method.
 
-            val = tb = Py_None;
-            exc = TOP();
-            if (exc == NULL) {
-                STACKADJ(-1);
-                exit_func = TOP();
-                SET_TOP(exc);
-                exc = Py_None;
-            }
-            else {
-                assert(PyExceptionClass_Check(exc));
-                PyObject *tp2, *exc2, *tb2;
-                PyTryBlock *block;
-                val = SECOND();
-                tb = THIRD();
-                tp2 = FOURTH();
-                exc2 = PEEK(5);
-                tb2 = PEEK(6);
-                exit_func = PEEK(7);
-                SET_VALUE(7, tb2);
-                SET_VALUE(6, exc2);
-                SET_VALUE(5, tp2);
-                /* UNWIND_EXCEPT_HANDLER will pop this off. */
-                SET_FOURTH(NULL);
-                /* We just shifted the stack down, so we have
-                   to tell the except handler block that the
-                   values are lower than it expects. */
-                assert(f->f_iblock > 0);
-                block = &f->f_blockstack[f->f_iblock - 1];
-                assert(block->b_type == EXCEPT_HANDLER);
-                assert(block->b_level > 0);
-                block->b_level--;
-            }
-
-            stack[0] = exc;
-            stack[1] = val;
-            stack[2] = tb;
-            res = _PyObject_FastCall(exit_func, stack, 3);
+               We shift the top 6 values of the stack down and push the result
+               of the call EXIT(TOP, SECOND, THIRD).
+            */
+            PyObject *exc = TOP();
+            assert(PyExceptionClass_Check(exc));
+            PyObject *val = SECOND();
+            PyObject *tb = THIRD();
+            PyObject *tp2 = FOURTH();
+            PyObject *exc2 = PEEK(5);
+            PyObject *tb2 = PEEK(6);
+            PyObject *exit_func = PEEK(7);
+            SET_VALUE(7, tb2);
+            SET_VALUE(6, exc2);
+            SET_VALUE(5, tp2);
+            SET_FOURTH(tb);
+            SET_THIRD(val);
+            SET_SECOND(exc);
+            /* We just shifted the stack down, so we have
```

Comment:
Hmm... instead of fixing the except handler block, can't we put a NULL or None at the seventh stack slot? That would feel less hackish.

Reply:
Interesting idea. I'll try to implement it.

While I thought about it I have found a bug in existing code (see bpo-33026).

Reply:
Done.